tests/test_ollama.py

```diff
@@ -0,0 +1,249 @@
+if not enable_ollama_tests or enable_ollama_tests == "0":
+    pytest.skip(allow_module_level=True)
+
+
+@pytest.fixture()
+def ollama_host() -> str:
+    ollama_host = os.environ["OLLAMA_HOST"]
+    return ollama_host
+
+
+@pytest.fixture()
+def con(db_url) -> psycopg.Connection:
+    return psycopg.connect(db_url)
+
+
+@pytest.fixture()
+def cur(con) -> psycopg.Cursor:
+    with con:
+        with con.cursor() as cursor:
+            yield cursor
+
+
+@pytest.fixture()
+def cur_with_ollama_host(ollama_host, con) -> psycopg.Cursor:
+    with con:
+        with con.cursor() as cursor:
+            cursor.execute(
+                "select set_config('ai.ollama_host', %s, false) is not null",
+                (ollama_host,),
+            )
+            yield cursor
+
+
+def test_ollama_list_models_no_host(cur_with_ollama_host):
+    cur_with_ollama_host.execute("select count(*) > 0 from ai.ollama_list_models()")
+    actual = cur_with_ollama_host.fetchone()[0]
+    assert actual is True
+
+
+def test_ollama_list_models(cur, ollama_host):
+    cur.execute(
+        "select count(*) > 0 from ai.ollama_list_models(_host=>%s)", (ollama_host,)
+    )
+    actual = cur.fetchone()[0]
+    assert actual is True
+
+
+def test_ollama_embed(cur, ollama_host):
+    cur.execute(
+        """
+        select vector_dims
+        (
+            ai.ollama_embed
+            ( 'llama3'
+            , 'the purple elephant sits on a red mushroom'
+            , _host=>%s
+            )
+        )
+    """,
+        (ollama_host,),
+    )
+    actual = cur.fetchone()[0]
+    assert actual == 4096
+
+
+def test_ollama_embed_no_host(cur_with_ollama_host):
+    cur_with_ollama_host.execute("""
+        select vector_dims
+        (
+            ai.ollama_embed
+            ( 'llama3'
+            , 'the purple elephant sits on a red mushroom'
+            )
+        )
+    """)
+    actual = cur_with_ollama_host.fetchone()[0]
+    assert actual == 4096
+
+
+def test_ollama_embed_via_openai(cur, ollama_host):
+    cur.execute(
+        """
+        select vector_dims
+        (
+            ai.openai_embed
+            ( 'llama3'
+            , 'the purple elephant sits on a red mushroom'
+            , _api_key=>'this is a garbage api key'
+            , _base_url=>concat(%s::text, '/v1/')
+            )
+        )
+    """,
+        (ollama_host,),
+    )
+    actual = cur.fetchone()[0]
+    assert actual == 4096
+
+
+def test_ollama_generate(cur, ollama_host):
+    cur.execute(
+        """
+        select ai.ollama_generate
+        ( 'llama3'
+        , 'what is the typical weather like in Alabama in June'
+        , _system=>'you are a helpful assistant'
+        , _host=>%s
+        , _options=> jsonb_build_object
+          ( 'seed', 42
+          , 'temperature', 0.6
+          )
+        )
+    """,
+        (ollama_host,),
+    )
+    actual = cur.fetchone()[0]
+    assert "response" in actual and "done" in actual and actual["done"] is True
```

Comment:
Small thing. If one of these conditions fails, would it be easy to know which one? I really don't know, but I suspect you'd get something like assertion fail . Maybe it's better to have one assert statement per check instead of using `and`.

Don't change them, but maybe for future tests it's better to split them.

Reply:
Well I tested this. You get something like

```
======================================================= FAILURES ======================
________________________________________________________ test_a _______________________

    def test_a():
        a = 42
        b = "asd"
>       assert a == 42 and "afs" in b
E       AssertionError: assert (42 == 42 and 'afs' in 'asd')

tests/test_lambda.py:15: AssertionError
=============================================== short test summary info ===============
FAILED tests/test_lambda.py::test_a - AssertionError: assert (42 == 42 and 'afs' in 'as
================================================== 1 failed in 0.06s ==================
```